File: lib/calc_score/src/calc.c
#include <TH/TH.h>
#include <stdio.h>

void find_scores(int K, int N, float thresh, int array_size, int *pos,int *pos_indices, float *actioness, float *overlaps_scr,
		 float *scores, float *overlaps, int indx, /* now the return arrays */
		 int *next_pos, int *next_pos_indices, float* next_actioness, float* next_overlaps_scr, float* f_scores){
/* void find_scores(int K, float thresh, int array_size, int *pos,int *pos_indices, float *actioness, float *overlaps_scr, */
/* 		 float *scores, float *overlaps, int indx, /\* now the return arrays *\/ */
/* 		 int *next_pos, int *next_pos_indices, float* next_actioness, float* next_overlaps_scr, float* f_scores){ */

  float tmp_sum;
  int idx_i, idx_j, m, tmp_pos;
  float **new_scores, **new_ovelaps;     // new_scores and new_overlaps are updated in every loop
  int next_pos_max_size;
  next_pos_max_size = array_size * K;

  // init arrays
  new_scores  = (float **) malloc (array_size * sizeof(float *));
  new_ovelaps = (float **) malloc (array_size * sizeof(float *));
  for (int j=0; j<array_size; j++){
    new_scores[j]  = (float *) malloc (K * sizeof (float));
    new_ovelaps[j] = (float *) malloc (K * sizeof (float));
  }

  // count sums for actioness and overlaps
  for (int j=0; j<array_size; j++){ // for each good tube

    idx_i = pos[j * N * 2 + pos_indices[j] * 2];
    idx_j = pos[j * N * 2 + pos_indices[j] * 2+ 1];

    for(int z=0; z<K; z++){
      new_scores[j][z] = actioness[j] + scores[z];
      new_ovelaps[j][z] = overlaps_scr[j] + overlaps[idx_i* K * K + idx_j * K + z];
    }

  }

  for (int j=0; j<array_size; j++){
    for (int z=0; z<K; z++){
      m = pos_indices[j]+1;
      tmp_sum = new_scores[j][z] / (m+1) + new_ovelaps[j][z] / m;
      tmp_pos = j * K + z;
      /* printf("tmp_pos : %d tmp_sum %f ", tmp_pos,tmp_sum); */
	
      if (tmp_sum > thresh){
	/* printf(" --> mpike\n"); */
	for (int q=0; q<pos_indices[j]+1; q++){
	  next_pos[tmp_pos * N * 2 + q * 2 + 0]=pos[j * N * 2 + q * 2 + 0];
	  next_pos[tmp_pos * N * 2 + q * 2 + 1]=pos[j * N * 2 + q * 2 + 1];
	}

	// update pos indices and add new tube
	next_pos_indices[tmp_pos]=pos_indices[j]+1;

	next_pos[tmp_pos * N * 2 + next_pos_indices[tmp_pos] * 2 + 0]=indx;
	next_pos[tmp_pos * N * 2 + next_pos_indices[tmp_pos] * 2 + 1]=z;

	next_actioness[tmp_pos] = new_scores[j][z];
	next_overlaps_scr[tmp_pos] = new_ovelaps[j][z];
	f_scores[tmp_pos] = tmp_sum;

      }
      else{
	/* printf("\n"); */
	next_pos_indices[tmp_pos] = -1;
	next_actioness[tmp_pos] = -1;
	next_overlaps_scr[tmp_pos] = -1;
	f_scores[tmp_pos] = -1;
      }
    }
  }
  // add now time for adding
  tmp_pos++;

  for (int z=0; z<K; z++){
    next_pos_indices[tmp_pos]=0;
    next_pos[tmp_pos * N * 2    ]=indx;
    next_pos[tmp_pos * N * 2 + 1]=z;
    next_actioness[tmp_pos] = scores[z];
    next_overlaps_scr[tmp_pos] = 0;
    f_scores[tmp_pos] = scores[z];

    tmp_pos++;
  }
}

int calc_test(int K, int N, float thresh, int array_size, THIntTensor *pos, THIntTensor *pos_indices, THFloatTensor * actioness,
	      THFloatTensor * overlaps_scr, THFloatTensor * scores, THFloatTensor * overlaps, int idx,  /* return arrays */
	      THIntTensor * next_pos, THIntTensor * next_pos_indices, THFloatTensor * next_actioness,
	      THFloatTensor * next_overlaps_scr, THFloatTensor * f_scores)
{
  int   * pos_data               = THIntTensor_data(pos);
  int   * pos_indices_data       = THIntTensor_data(pos_indices);

  float * actioness_data         = THFloatTensor_data(actioness);
  float * overlaps_scr_data      = THFloatTensor_data(overlaps_scr);

  float * scores_data            = THFloatTensor_data(scores);
  float * overlaps_data          = THFloatTensor_data(overlaps);

  float * next_pos_data          = THIntTensor_data(next_pos);
  float * next_pos_indices_data  = THIntTensor_data(next_pos_indices);
  float * next_actioness_data    = THFloatTensor_data(next_actioness);
  float * next_overlaps_scr_data = THFloatTensor_data(next_overlaps_scr);
  float * f_scores_data          = THFloatTensor_data(f_scores);

  /* printf("ok until here...\n"); */
  /* printf("pos_data[0]: %d\n",pos_data[0]); */
  /* printf("pos_data[1]: %d\n",pos_data[1]); */
  /* printf("pos_data[2]: %d\n",pos_data[2]); */
  /* printf("pos_data[3]: %d\n",pos_data[3]); */
  /* printf("pos_data[4]: %d\n",pos_data[4]); */
  /* printf("pos_data[5]: %d\n",pos_data[5]); */
  find_scores(K, N, thresh, array_size, pos_data, pos_indices_data, actioness_data, overlaps_scr_data,
  	      scores_data, overlaps_data, idx,
  	      next_pos_data, next_pos_indices_data, next_actioness_data, next_overlaps_scr_data, f_scores_data);

  return 1;
}

File: lib/calc_score/calc.py
import torch
from torch.autograd import Function
# from _ext import calc as c
from ._ext import calc as c

import sys
import numpy

class Calculator(Function):

    def __init__(self, k, update_thresh, thresh, ):
        self.k = k
        self.update_thresh = update_thresh
        self.thresh = thresh

    def forward(self, N, K, array_size, pos, pos_indices, \
                actioness, overlaps_scr, overlaps, scores, indx):
        # print('self.thresh :',self.thresh)
        N = N.int().item()
        K = K.int().item()
        array_size = array_size.int().item()
        indx = indx.int().item()

        zeros_t = torch.zeros(K,N,2).int().cuda()-1
        # for next
        next_pos_max_size = array_size * K + K
        next_pos = pos.new(next_pos_max_size, N, 2).zero_().view(-1) -1
        next_pos_indices = pos_indices.new(next_pos_max_size).zero_() -1
        next_actioness = actioness.new(next_pos_max_size).zero_()
        next_overlaps_scr = overlaps_scr.new(next_pos_max_size).zero_()
        f_scores = actioness.new(next_pos_max_size).zero_()
        # print('overlaps.shape :',overlaps.type())
        # print('pos.type() :',pos.type())
        # print('pos_indices.type() :',pos_indices.type())
        # print('actioness.type() :',actioness.type())
        # print('overlaps_scr.type() :',overlaps_scr.type())
        # print('scores.type() :',scores.type())
        # print('next_pos.type() :',next_pos.type())
        # print('next_pos_indices.type() :',next_pos_indices.type())
        # print('next_actioness.type() :',next_actioness.type())
        # print('next_overlaps_scr.type() :',next_overlaps_scr.type())
        # print('f_scores.type() :',f_scores.type())
        # print('overlaps.shape :',overlaps.shape)
        # print('pos.type() :',pos.shape)
        # print('pos_indices.type() :',pos_indices.shape)
        # print('actioness.type() :',actioness.shape)
        # print('overlaps_scr.type() :',overlaps_scr.shape)
        # print('scores.type() :',scores.shape)

        c.calc_test_cuda(K, N, self.thresh, array_size, pos.view(-1), pos_indices, actioness,
                         overlaps_scr, scores, overlaps.view(-1), 0,
                         next_pos, next_pos_indices, next_actioness,
                         next_overlaps_scr, f_scores)

        # print('next_overlaps_scr :',next_overlaps_scr)
        next_pos = next_pos.view(next_pos_max_size, N, 2)
        for i in range(next_pos_indices.size(0)):
            if next_pos_indices[i] > -1:
                z = i // K
                for j in range(next_pos_indices[i]):
                    next_pos[i,j,0] = pos[z,j,0]
                    next_pos[i,j,1] = pos[z,j,1]
                next_pos[i,next_pos_indices[i],0] = indx
                next_pos[i,next_pos_indices[i],1] = i % K

        over_thresh_idx = next_pos_indices.gt(-1).nonzero().squeeze()

        # if over_thresh_idx.nelement() == 0:
        #     print('Empty this loop..., self thresh: ', self.thresh, ' pos.shape :',pos.shape)

        next_pos = next_pos[over_thresh_idx]
        next_pos_indices =  next_pos_indices[over_thresh_idx]

        f_scores = f_scores[over_thresh_idx]
        next_actioness = next_actioness[over_thresh_idx]
        next_overlaps_scr = next_overlaps_scr[over_thresh_idx]

        return next_pos, next_pos_indices, f_scores, next_actioness, next_overlaps_scr

    def update_scores(self, final_scores, final_poss, f_scores, pos, pos_indices, actioness, \
                      overlaps_scr):

        # print('Updating thresh')
        ## first update next loop

        _, indices = torch.sort(f_scores,descending=True)

        if self.thresh == f_scores[indices[self.k]].item():

            print('f_scores[:self.k] :',f_scores[:self.k].cpu().numpy())
            self.thresh = self.thresh + 0.001
        else:
            self.thresh = f_scores[indices[self.k]].item()

        indices = indices[:self.k]
        self.thresh = f_scores[self.k].item()

        pos = pos[indices].contiguous()
        pos_indices = pos_indices[indices].contiguous()
        actioness = actioness[indices].contiguous()
        overlaps_scr = overlaps_scr[indices].contiguous()
        f_scores = f_scores[indices].contiguous()

        ## now time for final scores
        indices = final_scores.ge(self.thresh).nonzero()

        indices = indices.squeeze()
        # print('final_scores.shape :',final_scores.shape)
        # print('final_poss.shape :',final_poss.shape)
        final_scores = final_scores[indices].contiguous()
        final_poss = final_poss[indices].contiguous()

        return final_scores, final_poss, pos, pos_indices, \
            actioness, overlaps_scr,  f_scores
        
if __name__ == '__main__':

    # with open('../overlaps.pwf', 'rb') as fp:
    overlaps = torch.load('../overlaps.pwf')
    scores = torch.load('../scores.pwf')
    print('overlaps.shape :',overlaps.shape)
    print('scores.shape :',scores.shape)
    # scores = torch.rand(5,20).cuda()
    # overlaps = torch.rand(4,20,20).cuda()
    N = torch.Tensor([5])
    K = torch.Tensor([20])
    calc = Calculator(100, 1000, 1.5)
    f = calc(overlaps, scores, N,K)
    
